[204] Optimal clustering parameters

Jack-0-0 opened the pull request:
Closes #204.

This PR:

- Updates the `Vectors` class to initialise with no vectors, removes some of the repetition in the init function, simplifies `add_vectors`.
- Splits clustering function `hdbscan_clustering` into dimension reduction and clustering function.
- Adds function `kmeans_clustering` to cluster using K-Means.
- Adds functions `kmeans_param_grid_search` and `hdbscan_param_grid_search` to search clustering parameters and find related mean silhouette scores.
- Adds a notebook showing example usage (`analysis/examples/11_cluster_optimisation_gtr.py`)

---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

Jack-0-0 committed:
Remove repitition from Vectors init and fix to wor with no vectors

Jack-0-0 committed:
Remove print statement

Jack-0-0 committed:
Simplify add_vectors

Jack-0-0 committed:
Remove unused variable

Jack-0-0 committed:
Add kmeans and grid search functions

Jack-0-0 committed:
Add bool for hdbscan clustering

Jack-0-0 committed:
Add highest_silhouette_model_params

Jack-0-0 committed:
Add notebook that uses new cluster functions

beingkk commented on innovation_sweet_spots/utils/cluster_analysis_utils.py:
```diff
@@ -35,35 +43,102 @@
+            silhouette = silhouette_score(reduced_dims_vectors, clusters.labels.values)
+            method_record.append("HDBSCAN")
+        else:
+            clusters = kmeans_clustering(reduced_dims_vectors, kmeans_params=parameters)
+            silhouette = silhouette_score(reduced_dims_vectors, clusters)
```
Perhaps we can use a pre-computed distance matrix here instead of `reduced_dims_vectors` to speed up the grid search (for `silhouette_score` )

Jack-0-0 replied:
I also noticed that the umap reducer was being run on each parameter search iteration.
I ran some experiments using precomputed distances and running the reducer only once.
For HDBSCAN it reduces each iteration from ~22 secs down to ~4 secs.
For K-Means it reduces each iteration from ~20 secs down to ~2 secs.

Jack-0-0 committed:
Speed up param_grid_search

beingkk committed:
added a visualisation of the optimal cluster

beingkk committed:
added a visualisation of the optimal cluster

beingkk commented on innovation_sweet_spots/utils/cluster_analysis_utils.py:
```diff
@@ -35,35 +43,107 @@
+    clusterer.fit(vectors)
     cluster_probs = hdbscan.all_points_membership_vectors(clusterer)
-    best_cluster_prob = np.array([(np.argmax(x), np.max(x)) for x in cluster_probs])
+    highest_cluster_prob = np.array([(np.argmax(x), np.max(x)) for x in cluster_probs])
```
Could we adjust this line so that the user can choose whether to use the "-1" (noise) cluster label, or rather instead ignore it and assign the points to the next best cluster? Otherwise, I think it assigns a lot of points to the noise cluster, which probably will impact the silhouette score...

beingkk replied:
I think by default we should ignore the noise cluster

beingkk reviewed (changes requested):
Hey @Jack-0-0 thanks so much - it looks very good to me! I've added a few lines that also allows to visualise a clustering result using the best parameters (so, you should `git pull` to see these changes)

I only have two requests, which I hope are not too tricky to implement - do let me know if you think the time is too short:

- Given that the clustering is rather quick, could you please add an option to run each parameter set `n` times using different random seeds, where `n` can be chosen by the user (eg, default could be `n=1` but the user could also set it to higher in which case the function could return mean and standard deviation of the silhouette score). In that way, we'll have increased confidence that the silhouette score for the optimal parameter set is robust

- For the hdbscan clustering results, could you please adjust the `hdbscan_clustering` function so that it doesn't use the "-1" cluster (which are the points that are considered noisy). I think this will otherwise have an impact on the silhouette score. I’ve made a corresponding comment in the code - we can also discuss during tomorrow’s catchup.

Jack-0-0 committed:
Have separate grid search functions for Kmeans and HDBSCAN and add random state

Jack-0-0 committed:
Update example notebook to use new functions

Jack-0-0 commented:
It seems that HDBSCAN is deterministic. I have therefore split the `param_grid_search` function into two functions, `kmeans_param_grid_search` and `hdbscan_param_grid_search`.

`kmeans_param_grid_search` has an additional parameter `random_seeds` which is a list of random seeds. The function will do paramer search for each seed in `random_seeds` and then take the mean of all the random seed runs.

I have removed the random seed parameter for `hdbscan_clustering`.

Jack-0-0 commented:
Regarding not using the noise labels, the way the function was originally implemented, it wasn't using the noise labels but was assigning a label that corresponds to the index of the largest value in `hdbscan.all_points_membership_vectors` output (of which none relate to noise).

However, I noticed that in some cases the probabilities in `hdbscan.all_points_membership_vectors` for some vectors are all 0s. So in that case, I have changed it so the vector will be assigned a `-2` label.

I have added a way to have some of the points be labelled as noise using `labels_` built into HDBSCAN. To allow points to be assigned to noise, set `have_noise_labels=True`.

Interestingly, when allowing points to be assigned to noise, the average silhouette scores go negative.

Jack-0-0 committed:
Update functions to allow for noise labels

Jack-0-0 committed:
Update example notebook to use have_noise_labels parameter

beingkk committed:
Merge branch 'dev' into 204_cluster

beingkk commented:
Thank you!
